dopamine/discrete_domains/run_experiment.py

```diff
@@ -555,10 +576,36 @@ def __init__(
         environment, and returns an agent.
       create_environment_fn: A function which receives a problem name and
         creates a Gym environment for that problem (e.g. an Atari 2600 game).
+      max_simultaneous_iterations: int, maximum number of iterations running
+        simultaneously in separate threads.
       **kwargs: Additional positional arguments.
     """
     threading_utils.initialize_local_attributes(
         self, _environment=create_environment_fn)
+    self._running_iterations = threading.Semaphore(max_simultaneous_iterations)
+    self._output_lock = threading.Lock()
+
     super(AsyncRunner, self).__init__(
         base_dir=base_dir, create_agent_fn=create_agent_fn,
         create_environment_fn=create_environment_fn, **kwargs)
+
+  # TODO(aarg): Decouple experience generation from training.
+  def _run_experiment_loop(self):
+    """Runs iterations in multiple threads until `num_iterations` is reached."""
+    threads = []
+    for iteration in range(self._start_iteration, self._num_iterations):
+      self._running_iterations.acquire()
+      threads.append(self._run_one_iteration(iteration))
+    # Wait for all running iterations to complete.
+    for thread in threads:
+      thread.join()
+
+  @async_method
+  def _run_one_iteration(self, iteration):
+    """Runs one iteration in separate thread, logs and checkpoints results."""
```

Comment:
It means the exact same thing. Except that instead of returning the summary statistics they are directly logged to log files and Tensorboard.

Updated the docstring to make this clearer

Reply:
I'm still not sure that iteration is well defined.  I don't think it's your intention that it remains N environment samples, N training steps, but rather becomes N environment samples and... I'm actually not sure how training enters into the picture.

Reply:
Ok -- now that we've talked over the design, I understand what an iteration is.

That it was difficult for me to understand is a sign that there could be some more documentation.  But now that I do understand it, I don't see any places where I think that documentation is needed.  Maybe, then, optionally cite `Runner._run_one_iteration` or its docstring?  I like how clearly it's defined there.

Reply:
Added a quote in the class docstring.